driver-async/src/main/com/mongodb/async/client/gridfs/GridFSBucket.java

```diff
@@ -0,0 +1,294 @@
+import com.mongodb.WriteConcern;
+import com.mongodb.annotations.ThreadSafe;
+import com.mongodb.async.SingleResultCallback;
+import com.mongodb.client.gridfs.model.GridFSDownloadByNameOptions;
+import com.mongodb.client.gridfs.model.GridFSUploadOptions;
+import org.bson.BsonValue;
+import org.bson.conversions.Bson;
+import org.bson.types.ObjectId;
+
+/**
+ * Represents a GridFS Bucket
+ *
+ * @since 3.3
+ */
+@ThreadSafe
+public interface GridFSBucket {
+
+    /**
+     * The bucket name.
+     *
+     * @return the bucket name
+     */
+    String getBucketName();
+
+    /**
+     * Sets the chunk size in bytes. Defaults to 255.
+     *
+     * @return the chunk size in bytes.
+     */
+    int getChunkSizeBytes();
+
+    /**
+     * Get the write concern for the GridFSBucket.
+     *
+     * @return the {@link com.mongodb.WriteConcern}
+     */
+    WriteConcern getWriteConcern();
+
+    /**
+     * Get the read preference for the GridFSBucket.
+     *
+     * @return the {@link com.mongodb.ReadPreference}
+     */
+    ReadPreference getReadPreference();
+
+    /**
+     * Get the read concern for the GridFSBucket.
+     *
+     * @return the {@link com.mongodb.ReadConcern}
+     * @mongodb.server.release 3.2
+     * @mongodb.driver.manual reference/readConcern/ Read Concern
+     */
+    ReadConcern getReadConcern();
+
+    /**
+     * Create a new GridFSBucket instance with a new chunk size in bytes.
+     *
+     * @param chunkSizeBytes the new chunk size in bytes.
+     * @return a new GridFSBucket instance with the different chunk size in bytes
+     */
+    GridFSBucket withChunkSizeBytes(int chunkSizeBytes);
+
+    /**
+     * Create a new GridFSBucket instance with a different read preference.
+     *
+     * @param readPreference the new {@link ReadPreference} for the database
+     * @return a new GridFSBucket instance with the different readPreference
+     */
+    GridFSBucket withReadPreference(ReadPreference readPreference);
+
+    /**
+     * Create a new GridFSBucket instance with a different write concern.
+     *
+     * @param writeConcern the new {@link WriteConcern} for the database
+     * @return a new GridFSBucket instance with the different writeConcern
+     */
+    GridFSBucket withWriteConcern(WriteConcern writeConcern);
+
+    /**
+     * Create a new MongoDatabase instance with a different read concern.
+     *
+     * @param readConcern the new {@link ReadConcern} for the database
+     * @return a new GridFSBucket instance with the different ReadConcern
+     * @mongodb.server.release 3.2
+     * @since 3.3
+     * @mongodb.driver.manual reference/readConcern/ Read Concern
+     */
+    GridFSBucket withReadConcern(ReadConcern readConcern);
+
+    /**
+     * Opens a Stream that the application can write the contents of the file to.
+     * <p>
+     * As the application writes the contents to the returned Stream, the contents are uploaded as chunks in the chunks collection. When
```

Comment:
"Stream" should be lower case, since it doesn't match the return type of the method.